mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs

```diff
@@ -4356,6 +4544,223 @@ public static long pwritev (int fd, Iovec[] iov, long offset)
+
+		[DllImport (MPH, SetLastError=true, 
+				EntryPoint="Mono_Posix_Syscall_setsockopt_timeval")]
+		static extern unsafe int sys_setsockopt_timeval (int socket, int level, int option_name, ref Timeval option_value);
+
+		[DllImport (MPH, SetLastError=true, 
+				EntryPoint="Mono_Posix_Syscall_setsockopt_linger")]
+		static extern unsafe int sys_setsockopt_linger (int socket, int level, int option_name, ref Linger option_value);
+
+		public static unsafe int setsockopt (int socket, UnixSocketProtocol level, UnixSocketOptionName option_name, void *option_value, long option_len)
+		{
+			var _level = NativeConvert.FromUnixSocketProtocol (level);
+			var _option_name = NativeConvert.FromUnixSocketOptionName (option_name);
+			return sys_setsockopt (socket, _level, _option_name, option_value, option_len);
+		}
+
+		public static unsafe int setsockopt (int socket, UnixSocketProtocol level, UnixSocketOptionName option_name, int option_value)
+		{
+			return setsockopt (socket, level, option_name, &option_value, sizeof (int));
+		}
+
+		public static unsafe int setsockopt (int socket, UnixSocketProtocol level, UnixSocketOptionName option_name, byte[] option_value, long option_len)
+		{
+			if (option_len > (option_value == null ? 0 : option_value.Length))
+				throw new ArgumentOutOfRangeException ("option_len", "option_len > (option_value == null ? 0 : option_value.Length)");
+			fixed (byte* ptr = option_value)
+				return setsockopt (socket, level, option_name, ptr, option_len);
+		}
+
+		public static unsafe int setsockopt (int socket, UnixSocketProtocol level, UnixSocketOptionName option_name, Timeval option_value)
+		{
+			var _level = NativeConvert.FromUnixSocketProtocol (level);
+			var _option_name = NativeConvert.FromUnixSocketOptionName (option_name);
+			return sys_setsockopt_timeval (socket, _level, _option_name, ref option_value);
+		}
+
+		public static unsafe int setsockopt (int socket, UnixSocketProtocol level, UnixSocketOptionName option_name, Linger option_value)
+		{
+			var _level = NativeConvert.FromUnixSocketProtocol (level);
+			var _option_name = NativeConvert.FromUnixSocketOptionName (option_name);
+			return sys_setsockopt_linger (socket, _level, _option_name, ref option_value);
+		}
+
+		// shutdown(2)
+		//    int shutdown(int sockfd, int how);
+		[DllImport (LIBC, SetLastError=true, 
+				EntryPoint="shutdown")]
+		static extern int sys_shutdown (int socket, int how);
+
+		public static int shutdown (int socket, ShutdownOption how)
+		{
+			var _how = NativeConvert.FromShutdownOption (how);
+			return sys_shutdown (socket, _how);
+		}
+
+		// recv(2)
+		//    ssize_t recv(int sockfd, void *buf, size_t len, int flags);
+		[DllImport (MPH, SetLastError=true, 
+				EntryPoint="Mono_Posix_Syscall_recv")]
+		static extern unsafe long sys_recv (int socket, void *buffer, ulong length, int flags);
+
+		public static unsafe long recv (int socket, void *buffer, ulong length, MessageFlags flags)
```

Comment:
I've added such overloads for set/getsockopt(), recv() and send() and updated the pull request.